### ISSUE TITLE
Expecting private methods should never invoke them in Mockito

### ISSUE THREAD

**Issue by johanhaleby** [jayway/powermock#243]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on January 09, 2010 15:59:04_

A private method should of course never be invoked when expected.

_Original issue: http://code.google.com/p/powermock/issues/detail?id=223_

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on January 09, 2010 06:59:45_

public class PartialMockingExample {

	public String methodToTest() {
		return methodToMock();
	}

	public String methodToMock() {
		System.out.println("ikk");
		return "REAL VALUE";
	}
}

"methodToMock" should never print "ikk" which it does

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on March 10, 2010 05:28:30_

It works the same way in standard Mockito.

**Status:** Invalid

**Comment by johanhaleby** [comment]
_From [masph...@gmail.com](https://code.google.com/u/106133770804855118330/) on March 22, 2012 08:20:01_

How can this be invalid? If you want to mock a private method to get a well defined behaviour and avoid side effects when testing you don't want the mocked method to be executed! What if the private method accesses resources that are not available in your testing context? You'll most likely end up with null pointer exceptions or similar.

**Comment by johanhaleby** [comment]
_From [erichmusick](https://code.google.com/u/erichmusick/) on December 04, 2012 19:19:06_

For anyone who finds this 2+ years after the issue was closed, I wanted to provide a brief summary explaining how I've worked around this, since I didn't see the answer here, or on the original forum post that prompted the opening of this defect: https://groups.google.com/d/topic/powermock/cfZ4-xrWm5A/discussion As Johan mentioned, this is how standard Mockito works when using Spy... Recall, again, a sample from standard Mockito:

when(underTest.someMethod()).thenReturn(expected);

There is no option other than to call the actual method with this line. Reason? Order of operations. In order to set up this mock, the inner expression (underTest.someMethod()) is first evaluated. The mock hasn't yet been established, so the spy delegates to the actual class, because it doesn't know any better. Only after the actual method has completed its execution is 'when' called, which will establish the mock for all subsequent calls.

The workaround (again, standard Mockito):

doReturn(expected).when(underTest).someMethod();

Applying the same principle with a private method and PowerMockito, establishing the mock looks like this:

PowerMockito.doReturn(expected).when(underTest, nameOfMethodToMock, input1, input2);
